### PR TITLE
Permanently valid Queue.Item.id

### PR DESCRIPTION
**Pull request by tfennelly**
Potential precursor to a fix of [JENKINS-26445](https://issues.jenkins-ci.org/browse/JENKINS-26445).

I'm sure I have this all wrong (better ways etc) but I think the changes I made are enough to see what I'm "trying" to do i.e. somehow be able to get the original build queue `Item` id, from its subsequent `Run` instance, allowing us to view pending, running, completed etc builds as a sequence that can be paged up and down through as a single set (vs pending builds being different to everything else).
- (1b64ffb) First commit has changes to persist the actual build queue `COUNTER` so it can survive restarts.
- (0daec9f) Second commit introduces `QueueIDTrackerAction` as a way of providing access to the original queue id for a `Run` instance. This was a punt as to how "best" to do this in Jenkins land and is probably pants :).

With these changes (and fixes to them to get them right) we should be able to page up/down a build history, treating it as a continuous sequence.

### CODE REVIEW

**Comment by kohsuke** [comment]
My summary of the conversation with Tom.

* I wonder if deprecating integer `Queue.Item.id` and switching to UUID would be better and more robust
* a whole `Action` to store just one field doesn't bother me, but if Tom wants to put this directly in Run, that seems OK to me, too.
* If we need to provide global reverse lookup from UUID to Run, fingerprint database is a convenient way to do this.

**Comment by tfennelly** [comment]
Thanks @kohsuke.

Right, I'm not fond of the `id` and I think it makes a lot of sense to have a `UUID` that tracks a build from its inception (task going into the build queue) all the way to the end of its life (completed, aborted etc).

File: core/src/main/java/hudson/model/queue/QueueIDTrackerAction.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+ * An {@link hudson.model.Action} for tracking a {@link hudson.model.Run} back to
+ * it's original {@link hudson.model.Queue} identifier (from it's original
+ * queue {@link hudson.model.Queue.WaitingItem}).
+ * <p/>
+ * This provides a unique sequence ID spanning all builds of a {@link hudson.model.Job},
+ * no matter what state build (pending, running, completed etc).
+ *
+ * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
+ */
+public class QueueIDTrackerAction extends InvisibleAction {
```
Comment (jglick) [inline core/src/main/java/hudson/model/queue/QueueIDTrackerAction.java]: Overkill; simpler to add it directly to `Run` as a field.

**Comment by jglick** [comment]
UUID seems like overkill to me. What is wrong with a numeric `id`, other than it needing to be `long` rather than `int`?

File: core/src/main/java/hudson/model/Queue.java [hunk 2]
```diff
@@ -343,6 +343,14 @@ public void setSorter(QueueSorter sorter) {
     }
 
     /**
+     * Simple queue state persistence object.
+     */
+    public static class State {
```
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]: Why is this `public`?

File: core/src/main/java/hudson/model/queue/QueueIDTrackerAction.java (hunk 1 above)
Comment (tfennelly) [inline core/src/main/java/hudson/model/queue/QueueIDTrackerAction.java]: Hah I did it this way thinking this was the way you guys would want to do it (having seen you attach meta-data in this way in other places e.g. in Workflow). I'm obviously not good at reading minds yet :)

**Comment by tfennelly** [comment]
@jglick UUID seems easier to me in that it does not need to be managed/incremented, persisted etc. I'm also ok with a numeric ID, but I don't agree that it's in some way "easier" to work with. I would agree it's more human friendly though :)

File: core/src/main/java/hudson/model/Queue.java (hunk 2 above)
Comment (tfennelly) [inline core/src/main/java/hudson/model/Queue.java]: My assumption was that XStream would need access to it without messing with accessibility modifiers

File: core/src/main/java/hudson/model/Queue.java [hunk 3]
```diff
@@ -343,6 +343,14 @@ public void setSorter(QueueSorter sorter) {
...
+        public int counter;
+        public List<Object> items = new ArrayList<Object>();
```
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]: Should be `List<Task>`.

File: core/src/main/java/hudson/model/Queue.java [hunk 4]
```diff
@@ -377,6 +396,9 @@ public synchronized void load() {
                             if(item.task==null)
                                 continue;   // botched persistence. throw this one away
 
+                            // backward compatibility - in case the serialized queue state was
+                            // a List of items (as of old), then the maxId is the max of the items in
+                            // the persisted queue.
```
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]: This seems convoluted.

If you are loading in a `List<Task>` from the old `queue.xml`, then we have to fall back to inferring a `maxId` as here, and setting `COUNTER` to it—fine.

But if we are loading a `State` from the new `queue.xml`, which is the normal code path going forward, we will ultimately set `COUNTER` to `state.counter`. Going through `maxId` in that case is just confusing.

IOW it would be clearer to have something like

```java
List<Task> tasks;
if (unmarshaledObj instanceof State) {
    tasks = ((State) unmarshaledObj).items;
    COUNTER = ((State) unmarshaledObj).counter;
} else {
    int maxId = 0;
    tasks = new ArrayList<Task>();
    for (Object o : (List) unmarshaledObj) {
        if (o instanceof Task) {
            tasks.add((Task) o);
            maxId = Math.max(maxId, ((Task) o).id);
        }
    }
    COUNTER = maxId;
}
```

which separates the compatibility code into its own block for easy ignoring.

**Comment by tfennelly** [comment]
@jglick and why would the `id` need to change to a `long` ?

File: core/src/main/java/hudson/model/Queue.java [hunk 5]
```diff
@@ -1809,6 +1834,10 @@ public WaitingItem(Calendar timestamp, Task project, List<Action> actions) {
             this.timestamp = timestamp;
         }
 
+        public static int getCurrentCounterValue() {
+            return COUNTER.intValue();
+        }
```
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]: What is this for?

File: core/src/main/java/hudson/model/queue/WorkUnit.java [hunk 6]
```diff
@@ -85,6 +86,10 @@ public Executable getExecutable() {
     @Restricted(NoExternalUse.class)
     public void setExecutable(Executable executable) {
         this.executable = executable;
+        if (executable instanceof Run) {
+            // TODO: Not possible via the QueueListener it seems. Is there something else?
```
Comment (jglick) [inline core/src/main/java/hudson/model/queue/WorkUnit.java]: Possible via `onLeft`. But this is more direct.

File: test/src/test/java/hudson/model/QueueTest.java [hunk 7]
```diff
@@ -100,6 +105,16 @@
...
     @Rule public JenkinsRule r = new NodeProvisionerRule(-1, 0, 10);
 
+    private File queueFile;
+
+    @Before
+    public void resetQueueState() throws IOException {
+        queueFile = new File(r.jenkins.getRootDir(), "queue.xml");
+        XmlFile xmlFile = new XmlFile(Queue.XSTREAM, queueFile);
+        xmlFile.write(new Queue.State());
+        r.jenkins.getQueue().load();
```
Comment (jglick) [inline test/src/test/java/hudson/model/QueueTest.java]: Why is this necessary? Each test gets its own `Jenkins` and thus its own `Queue` anyway.

File: test/src/test/java/hudson/model/QueueTest.java [hunk 8]
```diff
@@ -152,6 +169,47 @@
...
+        // The counter state should be maintained.
+        assertEquals(1, Queue.WaitingItem.getCurrentCounterValue());
+    }
+
+    /**
+     * Make sure the queue can be reconstructed from a List queue.xml.
+     * Prior to the Queue.State class, the Queue items were just persisted as a List.
+     */
+    @Test
+    public void recover_from_leegacy_list() throws Exception {
```
Comment (jglick) [inline test/src/test/java/hudson/model/QueueTest.java]: Typo.

File: test/src/test/java/hudson/model/QueueTest.java [hunk 9]
```diff
@@ -152,6 +169,47 @@
...
+        // prevent execution to push stuff into the queue
+        r.jenkins.setNumExecutors(0);
+        r.jenkins.setNodes(r.jenkins.getNodes());
+
+        FreeStyleProject testProject = r.createFreeStyleProject("test");
+        testProject.scheduleBuild(new UserIdCause());
+
+        // Persist the List only (as before Queue.State)
+        XmlFile xmlFile = new XmlFile(Queue.XSTREAM, queueFile);
+        xmlFile.write(new ArrayList(Arrays.asList(q.getItems())));
```
Comment (jglick) [inline test/src/test/java/hudson/model/QueueTest.java]: More convincing to do this as a `@LocalData` test packaging an actual old `queue.xml` as determined experimentally.

File: core/src/main/java/hudson/model/queue/QueueIDTrackerAction.java [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+    public static final int QUEUE_ID_UNKNOWN = -1;
```
Comment (jglick) [inline core/src/main/java/hudson/model/queue/QueueIDTrackerAction.java]: BTW `0` would probably suffice, since it seems from reading the code that the first item is always given an ID of 1.

File: core/src/main/java/hudson/model/Queue.java (hunk 4 above)
Comment (tfennelly) [inline core/src/main/java/hudson/model/Queue.java]: So there was always a loop in there performing checks on the persisted list items ala

```
for (Object o : state.items) {
    if (o instanceof Task) {
        // backward compatibility
        schedule((Task)o, 0);
    } else if (o instanceof Item) {
        Item item = (Item)o;
        if(item.task==null)
            continue;   // botched persistence. throw this one away

        maxId = Math.max(maxId, item.id);
        if (item instanceof WaitingItem) {
            item.enter(this);
        } else if (item instanceof BlockedItem) {
            item.enter(this);
        } else if (item instanceof BuildableItem) {
            item.enter(this);
        } else {
            throw new IllegalStateException("Unknown item type! " + item);
        }
    } // this conveniently ignores null
}
```

i.e. there was more than the maxId tango going on. I intentionally did not touch that, other than to add the comment before `maxId = Math.max(maxId, item.id);`.

Sure the `maxId` stuff can be moved out of the loop, but I'd be for leaving the other code as is and leaving those checks as they were, whether we have a `State` or a `List`.

Comment (tfennelly) [inline core/src/main/java/hudson/model/Queue.java]: Of course if using a `UUID` there'd be none of the messing about with `maxId`s etc . Just saying :)

File: core/src/main/java/hudson/model/Queue.java (hunk 5 above)
Comment (tfennelly) [inline core/src/main/java/hudson/model/Queue.java]: Simply so it can be accessed and checked from a test. Could make it `protected`.

File: core/src/main/java/hudson/model/queue/WorkUnit.java (hunk 6 above)
Comment (tfennelly) [inline core/src/main/java/hudson/model/queue/WorkUnit.java]: You sure? I thought onLeft only had access to the `task` i.e. the project def. Seemed like `Executable` instance was not created at that stage.

File: test/src/test/java/hudson/model/QueueTest.java (hunk 7 above)
Comment (tfennelly) [inline test/src/test/java/hudson/model/QueueTest.java]: But not a new classloader. `Queue.WaitingItem.COUNTER` is a `private static`. Could change visibility, but ....

File: core/src/main/java/hudson/model/queue/QueueIDTrackerAction.java (hunk 10 above)
Comment (tfennelly) [inline core/src/main/java/hudson/model/queue/QueueIDTrackerAction.java]: Right, I just like setting that kind of thing to a value that the "thing" can never legitimately have at any stage.

File: core/src/main/java/hudson/model/queue/QueueIDTrackerAction.java (hunk 1 above)
Comment (jglick) [inline core/src/main/java/hudson/model/queue/QueueIDTrackerAction.java]: Metadata attached by a plugin _has_ to go through `Action` (typically `RunAction2`). In this case we expect every `Run` to have this metadata, and it is stored and retrieved by core code, so it might as well be a field, just like the `timestamp` and so on.

File: core/src/main/java/hudson/model/Queue.java (hunk 2 above)
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]: Nope, XStream deals with `private` types and fields, just like Java serialization.

File: core/src/main/java/hudson/model/Queue.java (hunk 4 above)
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]: For the compatibility loop you can look for `Task`, scheduling it; else look for `Item`, updating `maxId` and also adding it to the list of items to process. So

    List<Item> items;

is what I should have written.

File: core/src/main/java/hudson/model/Queue.java (hunk 5 above)
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]: Default access (no modifier) suffices.

File: core/src/main/java/hudson/model/queue/WorkUnit.java (hunk 6 above)
Comment (jglick) [inline core/src/main/java/hudson/model/queue/WorkUnit.java]: It gets a `LeftItem`, from which you can ask for an `Executable`.

File: test/src/test/java/hudson/model/QueueTest.java (hunk 7 above)
Comment (jglick) [inline test/src/test/java/hudson/model/QueueTest.java]: Then it ought to be made into an instance field of `Queue`. We do not want `static` state.

**Comment by jglick** [comment]
> UUID seems easier to me in that it does not need to be managed/incremented, persisted etc.

Well the incrementing is trivial. The persistence is a little work, but clearly not much. The benefit is easier debugging: IDs in functional tests will always go 1, 2, 3…, and you can more easily inspect a `queue.xml` to see what is going on because the `id`s will give a chronological order of scheduling.

Also if you switch to UUID you need to do something with `id`s of loaded `Item`s loaded from the last `queue.xml` before the switch. These need to be given UUIDs retroactively; but the integer `id` also still needs to be exported over the REST API for the benefit of clients listening for that item to be moved onto an executor; yet only these legacy `Item`s should have that old integer `id`, all new ones should display a UUID only.

Furthermore, if you switched `id` to be a `String` or `UUID`, some REST API clients expecting to have a number in this position might be broken. So this seems risky for compatibility. (In principle such clients could also be broken by a switch to `long`, but this is far less likely—they would have to have picked a 32-bit storage type to begin with, and the server would need to have scheduled 2³¹ builds.)

**Comment by jglick** [comment]
> why would the `id` need to change to a `long`?

Because if you are persisting the counter, eventually you might cycle around to negative numbers and then zero if you left it as `int`, causing bizarre problems. This might sound far-fetched, but consider that there was an `iota` that was used in Jelly page rendering, and we encountered support customers getting weird errors after that cycled through 2³² times! (A good problem for any tool to have.) Most likely `iota` would be incremented a couple of orders of magnitude more frequently than new builds are scheduled, but we are still within the realm of possibility here.

File: core/src/main/java/hudson/model/queue/WorkUnit.java (hunk 6 above)
Comment (jglick) [inline core/src/main/java/hudson/model/queue/WorkUnit.java]: (But again since we are inside core code it is more straightforward to do it like you did.)

**Comment by tfennelly** [comment]
@jglick @kohsuke I looked at doing the UUID and, aside from Jesse's issues with it, it wouldn't work for us as a means to page through the Build History. We need a sequence such that if a build gets deleted then the paging still works.

**Comment by tfennelly** [comment]
@jglick I changed `Queue.Item.id` to a `long` (from an `int`). This will cause issues for anyone referencing it from outside Jenkins core. It's ugly and not what you'd want to do, but maybe we should deprecate `id` (leaving it as an `int`) and introduce a new `long` property on `Queue.Item`?

**Comment by daniel-beck** [comment]
> We need a sequence such that if a build gets deleted then the paging still works.

You realize that submitting a build to the queue first does not mean it'll get the earlier build number, right?

**Comment by tfennelly** [comment]
@daniel-beck right, builds added to the queue at a later time may start building earlier.

**Comment by tfennelly** [comment]
@reviewbybees again please

File: core/src/main/java/hudson/model/Queue.java [hunk 11]
```diff
@@ -343,6 +343,14 @@ public void setSorter(QueueSorter sorter) {
     }
 
     /**
+     * Simple queue state persistence object.
+     */
+    static class State {
+        public int counter;
```
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]: Not `long`? (Ditto `COUNTER`.)

File: core/src/main/java/hudson/model/Run.java [hunk 12]
```diff
@@ -149,6 +154,12 @@
     public transient /*final*/ int number;
 
     /**
+     * The original Queue task ID from where this Run instance originated.
+     */
+    // TODO is 'queueId' an okay name for this? Feels like there might be something better.
```
Comment (jglick) [inline core/src/main/java/hudson/model/Run.java]: Seems OK to me.

File: core/src/main/java/hudson/model/Run.java [hunk 13]
```diff
@@ -403,6 +414,29 @@ public int compareTo(@Nonnull RunT that) {
...
+     */
+    @Exported
+    public long getQueueId() {
+        return queueId;
+    }
+
+    /**
+     * Set the queue item ID.
+     * <p/>
+     * Mapped from the {@link Queue.Item#id}. This method should never be called from outside Jenkins Core.
```
Comment (jglick) [inline core/src/main/java/hudson/model/Run.java]: > This method should never be called from outside Jenkins Core.

Redundant given the annotation.

**Comment by jglick** [comment]
Not sure if switching `int` to `long` can be made compatible here. You could try using `@AdaptField` so that the `int`-based accesses become method calls, then throw some kind of exception if the value is in fact above `MAX_INT`.

**Comment by tfennelly** [comment]
@jglick not sure I see how `@AdaptField` can stop the following code from breaking:

```
int itemQueueId = queueItem.id;
```

Where `queueItem` is obviously a `Queue.Item`.

Is there an example? I looked and didn't find.

**Comment by kohsuke** [comment]
I looked at the code more carefully after talking to @tfennelly offline.

 `@AdaptField`  can do field-to-field type coersion, or you can write the following`:

```
class Queue.Item {
    @AdaptField(was=int.class)
    public long id;
}
```

And any `GETFIELD int Queue.Item.id` becomes `GETFIELD long Queue.Item.id` followed by `L2I` for casting it back to int. It loses the digits, but I think it's OK.

If you really want to throw an exception beyond MAX_INT, we can define a compatibility getter that does that check.

**Comment by tfennelly** [comment]
I made that change adding `@AdaptField` as described by KK. I then created a test plugin that just had a QueueListener ala:

```
@Extension
public class MyQueueListener extends QueueListener {
    @Override
    public void onEnterWaiting(Queue.WaitingItem wi) {
        super.onEnterWaiting(wi);
        int queueId = wi.id;
        System.out.println("****** Queue.WaitingItem enter waiting: " + queueId);
    }
}
```

The plugin was built against jenkins 1.580.1 and worked fine on that. I then tried running the same plugin against a version of jenkins with the `long` `Queue.Item.id` (with `@AdaptField(was=int.class)`). The got the error shown below.

Even if this worked though, it seems like it will only work for code loaded from a plugin, right? What if someone created a lib (that tried to assign from `Queueu.Item.id`) and added it directly to the jenkins cp? I don't think that's caught by the bytecode `Transformer`, right?

```
INFO: Augmented all extensions
Feb 19, 2015 10:38:20 AM hudson.ExtensionFinder$Sezpoz scout
WARNING: Failed to scout org.jenkins.plugin.queueidtest.PluginImpl$MyQueueListener
java.lang.VerifyError: Expecting a stackmap frame at branch target 20
Exception Details:
  Location:
    org/jenkins/plugin/queueidtest/PluginImpl$MyQueueListener.onEnterWaiting(Lhudson/model/Queue$WaitingItem;)V @11: ifeq
  Reason:
    Expected stackmap frame at this location.
  Bytecode:
    0000000: 2a2b b700 182b 120b b800 1c99 0009 b400
    0000010: 20a7 0006 b400 233d b200 29bb 002b 59b7
    0000020: 002c 122e b600 321c b600 35b6 0039 b600
    0000030: 3fb1

	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:270)
	at hudson.ExtensionFinder$Sezpoz.scout(ExtensionFinder.java:680)
	at hudson.ClassicPluginStrategy.findComponents(ClassicPluginStrategy.java:365)
	at hudson.ExtensionList.load(ExtensionList.java:300)
	at hudson.ExtensionList.ensureLoaded(ExtensionList.java:253)
	at hudson.ExtensionList.iterator(ExtensionList.java:143)
	at hudson.ClassicPluginStrategy.findComponents(ClassicPluginStrategy.java:364)
	at hudson.ExtensionList.load(ExtensionList.java:300)
	at hudson.ExtensionList.ensureLoaded(ExtensionList.java:253)
	at hudson.ExtensionList.get(ExtensionList.java:158)
	at hudson.PluginManager$PluginUpdateMonitor.getInstance(PluginManager.java:1248)
	at hudson.maven.PluginImpl.init(PluginImpl.java:58)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at hudson.init.TaskMethodFinder.invoke(TaskMethodFinder.java:106)
	at hudson.init.TaskMethodFinder$TaskImpl.run(TaskMethodFinder.java:176)
	at org.jvnet.hudson.reactor.Reactor.runTask(Reactor.java:282)
	at jenkins.model.Jenkins$7.runTask(Jenkins.java:903)
	at org.jvnet.hudson.reactor.Reactor$2.run(Reactor.java:210)
	at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:744)

Feb 19, 2015 10:38:20 AM hudson.ExtensionFinder$GuiceFinder$SezpozModule configure
WARNING: Failed to load org.jenkins.plugin.queueidtest.PluginImpl$MyQueueListener
java.lang.VerifyError: Expecting a stackmap frame at branch target 20
Exception Details:
  Location:
    org/jenkins/plugin/queueidtest/PluginImpl$MyQueueListener.onEnterWaiting(Lhudson/model/Queue$WaitingItem;)V @11: ifeq
  Reason:
    Expected stackmap frame at this location.
  Bytecode:
    0000000: 2a2b b700 182b 120b b800 1c99 0009 b400
    0000010: 20a7 0006 b400 233d b200 29bb 002b 59b7
    0000020: 002c 122e b600 321c b600 35b6 0039 b600
    0000030: 3fb1

	at java.lang.Class.getDeclaredConstructors0(Native Method)
	at java.lang.Class.privateGetDeclaredConstructors(Class.java:2493)
	at java.lang.Class.getConstructors(Class.java:1517)
	at hudson.ExtensionFinder$GuiceFinder$SezpozModule.resolve(ExtensionFinder.java:482)
	at hudson.ExtensionFinder$GuiceFinder$SezpozModule.configure(ExtensionFinder.java:511)
	at com.google.inject.AbstractModule.configure(AbstractModule.java:62)
	at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:230)
	at com.google.inject.spi.Elements.getElements(Elements.java:103)
	at com.google.inject.internal.InjectorShell$Builder.build(InjectorShell.java:136)
	at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:104)
	at com.google.inject.Guice.createInjector(Guice.java:96)
	at com.google.inject.Guice.createInjector(Guice.java:73)
	at hudson.ExtensionFinder$GuiceFinder.<init>(ExtensionFinder.java:279)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
	at java.lang.Class.newInstance(Class.java:374)
	at net.java.sezpoz.IndexItem.instance(IndexItem.java:181)
	at hudson.ExtensionFinder$Sezpoz._find(ExtensionFinder.java:642)
	at hudson.ExtensionFinder$Sezpoz.find(ExtensionFinder.java:617)
	at hudson.ClassicPluginStrategy.findComponents(ClassicPluginStrategy.java:371)
	at hudson.ExtensionList.load(ExtensionList.java:300)
	at hudson.ExtensionList.ensureLoaded(ExtensionList.java:253)
	at hudson.ExtensionList.iterator(ExtensionList.java:143)
	at hudson.ClassicPluginStrategy.findComponents(ClassicPluginStrategy.java:364)
	at hudson.ExtensionList.load(ExtensionList.java:300)
	at hudson.ExtensionList.ensureLoaded(ExtensionList.java:253)
	at hudson.ExtensionList.get(ExtensionList.java:158)
	at hudson.PluginManager$PluginUpdateMonitor.getInstance(PluginManager.java:1248)
	at hudson.maven.PluginImpl.init(PluginImpl.java:58)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at hudson.init.TaskMethodFinder.invoke(TaskMethodFinder.java:106)
	at hudson.init.TaskMethodFinder$TaskImpl.run(TaskMethodFinder.java:176)
	at org.jvnet.hudson.reactor.Reactor.runTask(Reactor.java:282)
	at jenkins.model.Jenkins$7.runTask(Jenkins.java:903)
	at org.jvnet.hudson.reactor.Reactor$2.run(Reactor.java:210)
	at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:744)
```

**Comment by jglick** [comment]
My suggestion was a little different, something like

```java
class Queue.Item {
    private final long id;
    public long getId() {
        return id;
    }
    @AdaptField(was=int.class, name="id")
    @Deprecated
    public int getIdLegacy() {
        if (id > Integer.MAX_INT) {
            throw new IllegalStateException("update your baseline to deal with long-valued IDs");
        }
        return (int) id;
    }
}
```

Not sure exactly what `@AdaptField` supports; its Javadoc is very confusing.

**Comment by jglick** [comment]
> What if someone created a lib […] and added it directly to the jenkins cp?

Not a supported use case. (You _can_ create a library with `jenkins-module` packaging for some exotic purposes, but in this case you are on the hook to ensure compatibility with the current Jenkins core. Users cannot update such modules, only the WAR vendor.)

**Comment by tfennelly** [comment]
And your suggestion should work for legacy plugin code like in the `QueueListener` I showed earlier e.g.

```
int queueId = queueItem.id;
```

??

`@AdaptField` seems like a bit too much black magic to me tbh. My pref would to simply deprecate `Queue.Item.id` (leaving it as an `int`) and to introduce a new `long` version. I don't see it as any uglier than cryptic annotations, yet it's a lot clearer what's happening.

**Comment by jglick** [comment]
Yes, the idea is that that bytecode would get automatically converted to the equivalent of the source code

    int queueId = queueItem.getIdLegacy();

> deprecate `Queue.Item.id` (leaving it as an `int`) and to introduce a new `long` version

Possible; it lacks the runtime warning on overflow, but at least there is a compile-time deprecation. You would need to make the old field `transient` and restore it in `readResolve`.

**Comment by tfennelly** [comment]
@jglick that worked. Thanks. My test plugin called the injected method when trying to access `Queue.Item.id`. Cool (nearly too funky cool for my taste ;) ).

**Comment by jglick** [comment]
Well :+1: from me, nice work.

**Comment by tfennelly** [comment]
@jglick oops sorry ... I was thinking unsigned (not thinking enough) ... I'll fix that.

**Comment by tfennelly** [comment]
@reviewbybees hoping for another thumbs up

**Comment by stephenc** [comment]
:+1:

**Comment by jglick** [comment]
:+1: but for the future please avoid rebasing (“squashing”) commits as you just forced me to reread the whole diff for what I suspect was a one-line change (not to mention that we have lost all records of what people were reviewing before). Suffices to just push new commits to the same branch in a PR.

**Comment by tfennelly** [comment]
@jglick sorry Jesse, I thought we were done reviewing and that was why I waited 'til then to do the squashed. I think what you say about squash rebase makes sense when going through the review process, but I don't really agree with the idea of pushing a fragmented changeset upstream (for what is really a single change) Vs a single atomic commit that can be easily reviewed later in life.

File: core/src/main/java/hudson/model/Queue.java [hunk 14]
```diff
@@ -1480,8 +1509,24 @@ public Api getApi() {
          * VM-wide unique ID that tracks the {@link Task} as it moves through different stages
          * in the queue (each represented by different subtypes of {@link Item}.
          */
+	private final long id;
+
         @Exported
-    	public final int id;
+        public long getId() {
```
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]: Do not forget `@since`.

File: core/src/main/java/hudson/model/Run.java [hunk 15]
```diff
@@ -403,6 +413,29 @@ public int compareTo(@Nonnull RunT that) {
     }
 
     /**
+     * Get the {@link Queue.Item#getId()} of the original queue item from where this Run instance
+     * originated.
+     * @return The queue item ID.
+     * @since TODO
```
Comment (jglick) [inline core/src/main/java/hudson/model/Run.java]: Do not forget to update `@since`.

File: core/src/main/java/hudson/model/Run.java [hunk 16]
```diff
@@ -403,6 +413,29 @@ public int compareTo(@Nonnull RunT that) {
...
+    public long getQueueId() {
+        return queueId;
+    }
+
+    /**
+     * Set the queue item ID.
+     * <p/>
+     * Mapped from the {@link Queue.Item#getId()}.
+     * @param queueId The queue item ID.
+     * @since TODO
```
Comment (jglick) [inline core/src/main/java/hudson/model/Run.java]: No need for `@since` on a restricted method.

File: core/src/main/java/hudson/model/Run.java [hunk 17]
```diff
@@ -137,6 +137,11 @@
 public abstract class Run <JobT extends Job<JobT,RunT>,RunT extends Run<JobT,RunT>>
         extends Actionable implements ExtensionPoint, Comparable<RunT>, AccessControlled, PersistenceRoot, DescriptorByNameOwner, OnMaster {
 
+    /**
+     * The original {@link Queue.Item#getId()} has not yet been mapped onto the {@link Run} instance.
+     */
```
Comment (jglick) [inline core/src/main/java/hudson/model/Run.java]: `@since`

**Comment by jglick** [comment]
> a single atomic commit that can be easily reviewed later in life

That would be the merge commit; you can ignore the second parent chain.

**Comment by tfennelly** [comment]
@jglick ah ok... fair enough ... I learned something new there... thanks Jesse

File: core/src/main/java/hudson/model/Run.java [hunk 18]
```diff
@@ -403,6 +413,29 @@ public int compareTo(@Nonnull RunT that) {
...
+     * @return The queue item ID.
+     * @since TODO
+     */
+    @Exported
+    public long getQueueId() {
+        return queueId;
+    }
+
+    /**
+     * Set the queue item ID.
```
Comment (daniel-beck) [inline core/src/main/java/hudson/model/Run.java]: This Javadoc indicates the field name should probably be `queueItemId`.

File: core/src/main/java/hudson/model/Queue.java [hunk 19]
```diff
@@ -723,7 +752,7 @@ private synchronized WaitingItem peek() {
         return itemsView.get();
     }
 
-    public synchronized Item getItem(int id) {
```
Comment (daniel-beck) [inline core/src/main/java/hudson/model/Queue.java]: @tfennelly @jglick @stephenc 

This caused [JENKINS-27256](https://issues.jenkins-ci.org/browse/JENKINS-27256), breaking API access to individual queue items (`/queue/item/42/api/xml`), because Stapler does not support methods of the form `getX(long)`, only `getX(int)`.

See:
* http://stapler.kohsuke.org/reference.html *Public Getter Method with an int Argument*
* https://github.com/stapler/stapler/blob/master/core/src/main/java/org/kohsuke/stapler/MetaClass.java#L258...L275

Comment (tfennelly) [inline core/src/main/java/hudson/model/Queue.java]: Balls ... I did test this to make sure the `@AdaptField` still worked but this is a different issue that was not checked for (and was not expected).
